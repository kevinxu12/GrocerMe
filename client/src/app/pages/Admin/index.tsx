/**
 * @file Admin Dashboard
 * @author Kevin Xu
 */
import React from 'react';

/**
 * Render a dummy Admin Dashboard Component
 *
 * @returns {React.ReactElement} Admin Dashboard Component
 */
export const AdminDashboard = (): React.ReactElement => {
  return <div> Placeholder. Please make a dashboard eventually </div>;
};
