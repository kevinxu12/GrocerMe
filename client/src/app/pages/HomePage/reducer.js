/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
// const storagedata = JSON.parse(localStorage.getItem("data"));
export const initialState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        // Delete prefixed '@' from the github username
        draft.username = action.username;
        break;
    }
  });

export default homeReducer;
