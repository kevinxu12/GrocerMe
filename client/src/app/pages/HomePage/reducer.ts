/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { HomeAction } from 'types/actions';
import { HomeState } from 'types/RootState';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
export const initialState: HomeState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (
  state: HomeState = initialState,
  action: HomeAction,
): HomeState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        draft.username = action.payload.username;
        break;
    }
  });
};

export default homeReducer;
