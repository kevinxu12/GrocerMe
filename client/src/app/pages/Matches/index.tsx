/**
 * @file Frontend for Matches
 * @author Kevin Xu
 */

import {
  Card,
  CardActions,
  CardContent,
  Link,
  Stack,
  Typography,
} from '@mui/material';
import React, { useEffect, useState } from 'react';
import { ItemMatchRequest, RequestStatus } from 'types/rest';
import Api from 'utils/api';
import { Constants } from 'utils/constants';
import { parseAxiosSuccessResponse } from 'utils/request';
/**
 *
 */
export const Home = (): React.ReactElement => {
  const [matches, setMatches] = useState<ItemMatchRequest[]>([]);
  useEffect(() => {
    /**
     *
     */
    const fetchMatchesAsSupplier = async () => {
      const response = await Api.get('/itemMatchRequestsForUser', {
        status: RequestStatus.ACCEPTED,
      });
      const itemMatchRequests =
        parseAxiosSuccessResponse<ItemMatchRequest[]>(response);
      if (itemMatchRequests) {
        setMatches(itemMatchRequests);
      }
    };
    fetchMatchesAsSupplier();
  }, []);
  if (matches.length === 0) {
    return <div>No Matches for Now: Stay put</div>;
  }
  return (
    <div>
      <Stack spacing={2}>
        <Typography> All Matches as a Supplier </Typography>
        {matches.map((match: ItemMatchRequest, index: number) => (
          <Card key={index} sx={{ display: 'flex', width: '50%' }}>
            <CardContent>
              <Typography variant="body2" sx={{ mb: 1.5 }} component="div">
                Match between {match.requesterEmail} and {match.supplierEmail}
              </Typography>
            </CardContent>
            <CardActions>
              <Link href={Constants.MATCH + '/' + match._id}>View Match</Link>
            </CardActions>
          </Card>
        ))}
      </Stack>
    </div>
  );
};
