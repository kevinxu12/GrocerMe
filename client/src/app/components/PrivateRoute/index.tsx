/**
 * @file Route for loggedin users. If a user isn't logged in, should redirect back to landing page
 * @author Kevin Xu
 */
import React, { useState } from 'react';
import { Route, Redirect, useLocation } from 'react-router-dom';
import { isPathAuthenticated } from 'utils/auth';
import { Constants } from 'utils/constants';
import { initialState } from 'store/auth/reducer';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { Wrapper } from '../PrivateWrapper';

/**
 * @param {...any} root0 props to pass to component if authenticated
 * @param {React.ElementType} root0.component component to return if authenticated
 * @param {initialState} root0.auth the auth object containing username and roles
 * @returns {React.ElementType} either a Redirect or the component passed in props
 */
const PrivateRoute = ({
  component: Component,
  auth,
  ...rest
}): React.ReactElement => {
  const location = useLocation();
  const path = location.pathname;
  const [isAuthenticated, setIsAuthenticated] = useState(
    isPathAuthenticated(path, auth.roles),
  );
  useDeepCompareEffect(() => {
    console.log(path);
    setIsAuthenticated(isPathAuthenticated(path, auth.roles));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.roles, auth.username]);
  return (
    <Route
      {...rest}
      render={props => {
        return isAuthenticated ? (
          <Wrapper>
            <Component {...props} />
          </Wrapper>
        ) : (
          <Redirect to={Constants.HOME_URL} />
        );
      }}
    />
  );
};

export default PrivateRoute;
