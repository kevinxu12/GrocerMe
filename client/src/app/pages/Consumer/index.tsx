/**
 * @file A Home page for logged-in users, both consumers and suppliers
 * Dashboard is not a private route, because dashboard will refresh periodically to check if the user is still logged in or not.
 * @author Kevin Xu
 */
import React, { ComponentType, Dispatch, memo } from 'react';
import { AxiosResponse } from 'axios';
import changeAuth, { logout } from 'store/auth/actions';
import { Role, User } from 'types/rest';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { useHistory } from 'react-router-dom';
import { ChangeAuthAction, LogoutAction } from 'types/actions';
import { createStructuredSelector } from 'reselect';
import { makeSelectRoles, makeSelectUsername } from 'store/auth/selectors';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { Wrapper } from 'app/components/PrivateWrapper';
import { useSocket } from 'context/SocketContext';
import { Api, parseAxiosSuccessResponse } from 'utils/request';

interface DashboardPropsType {
  changeAuth: (email: string, roles?: Role[]) => void;
  logout: () => void;
  username: string;
  roles: Role[];
}

/**
 * Generate a dashboard React element. This should be the landing page for a logged-in, non-first-time user
 *
 * @param {object} props props passed
 * @param {Function} props.changeAuth check if logged in, if so store email into redux
 * @param {Function} props.logout logout the user from redux
 * @param {string} props.username the username of logged-in user in redux, if this exists
 * @param {Role[]} props.roles the roles of logged-in user in redux, if this exists
 * @returns {React.ReactElement} Dashboard Component
 */
const Dashboard = ({
  changeAuth,
  logout,
  username,
  roles,
}: DashboardPropsType): React.ReactElement => {
  const history = useHistory();
  const socket = useSocket();
  useDeepCompareEffect(() => {
    /**
     * Test function to listen to a socket thats emitted
     */
    async function checkLogin(): Promise<void> {
      const response: AxiosResponse = await Api.get('/loggedin');
      const user = parseAxiosSuccessResponse<User>(response);
      console.log(user);
      if (user == null) {
        logout();
        history.push('/');
        return;
      }
      changeAuth(user.email, user.roles);
      socket.emit('login', user);
    }
    checkLogin();
  }, [username, roles]);
  return <Wrapper>Sample Dashboard</Wrapper>;
};

/**
 * Maps redux store values to component props
 */
const mapStateToProps = createStructuredSelector({
  username: makeSelectUsername(),
  roles: makeSelectRoles(),
});

/**
 * Maps dispatch functions to component props
 *
 * @param {Dispatch} dispatch the dispatch object from redux
 * @returns {object} Object passed to props containing redux dispatch functions
 */
function mapDispatchToProps(
  dispatch: Dispatch<ChangeAuthAction | LogoutAction>,
) {
  return {
    /**
     * Stores info of logged-in user in redux
     *
     * @param {string} email the email of the logged-in user
     * @param {Role[]} roles the optional-roles of the logged-in user
     */
    changeAuth: (email: string, roles?: Role[]) => {
      dispatch(changeAuth({ username: email, roles }));
    },
    /**
     * Logs out a user if no email returned from redux (session object is expired)
     */
    logout: () => {
      dispatch(logout());
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose<ComponentType>(withConnect, memo)(Dashboard);
