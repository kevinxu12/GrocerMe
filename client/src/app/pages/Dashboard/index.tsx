/**
 * @file A Home page for logged-in users, both consumers and suppliers
 * @author Kevin Xu
 */
import React, { ComponentType, Dispatch, memo } from 'react';
import axios, { AxiosResponse } from 'axios';
import changeAuth, { logout } from 'store/auth/actions';
import { Role, StatusCode, User } from 'types/rest';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { useHistory } from 'react-router-dom';
import generateServerUrl from 'utils/url';
import { ChangeAuthAction, LogoutAction } from 'types/actions';
import { createStructuredSelector } from 'reselect';
import { makeSelectRoles, makeSelectUsername } from 'store/auth/selectors';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { Wrapper } from 'app/components/PrivateWrapper';

interface DashboardPropsType {
  changeAuth: (email: string, roles?: Role[]) => void;
  logout: () => void;
  username: string;
  roles: Role[];
}

/**
 * Generate a dashboard React element. This should be the landing page for a logged-in, non-first-time user
 *
 * @param {object} props props passed
 * @param {Function} props.changeAuth check if logged in, if so store email into redux
 * @param {Function} props.logout logout the user from redux
 * @param {string} props.username the username of logged-in user in redux, if this exists
 * @param {Role[]} props.roles the roles of logged-in user in redux, if this exists
 * @returns {React.ReactElement} Dashboard Component
 */
const Dashboard = (props: DashboardPropsType): React.ReactElement => {
  const history = useHistory();
  useDeepCompareEffect(() => {
    /**
     * Test function to listen to a socket thats emitted
     */
    async function checkLogin(): Promise<void> {
      const response: AxiosResponse = await axios.get(
        generateServerUrl('/loggedIn'),
        { withCredentials: true },
      );
      const responseBody: AxiosResponse['data'] = response.data;
      if (responseBody.statusCode === StatusCode.SUCCESS) {
        const user: User = responseBody.data;
        if (user) {
          console.log('changing props');
          props.changeAuth(user.email, user.roles);
        } else {
          props.logout();
          history.push('/');
        }
      }
    }
    checkLogin();
  }, [props.username, props.roles]);
  return <Wrapper>Sample Dashboard</Wrapper>;
};

/**
 * Maps redux store values to component props
 */
const mapStateToProps = createStructuredSelector({
  username: makeSelectUsername(),
  roles: makeSelectRoles(),
});

/**
 * Maps dispatch functions to component props
 *
 * @param {Dispatch} dispatch the dispatch object from redux
 * @returns {object} Object passed to props containing redux dispatch functions
 */
function mapDispatchToProps(
  dispatch: Dispatch<ChangeAuthAction | LogoutAction>,
) {
  return {
    /**
     * Stores info of logged-in user in redux
     *
     * @param {string} email the email of the logged-in user
     * @param {Role[]} roles the optional-roles of the logged-in user
     */
    changeAuth: (email: string, roles?: Role[]) => {
      dispatch(changeAuth({ username: email, roles }));
    },
    /**
     * Logs out a user if no email returned from redux (session object is expired)
     */
    logout: () => {
      dispatch(logout());
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose<ComponentType>(withConnect, memo)(Dashboard);
