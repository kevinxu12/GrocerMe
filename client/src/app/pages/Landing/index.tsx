/**
 * @file Component for HomePage
 * We shoud move the landing / home page out using Material UI instead of Hardcoded Components
 * @author Kevin Xu
 */
import React from 'react';
import styled from 'styled-components/macro';

import { PageWrapper } from 'app/components/PageWrapper';
import { StyledConstants } from 'styles/StyleConstants';
import OAuthButton from 'app/components/OAuthButton';
import { Constants } from 'utils/constants';

/**
 * @returns {React.FC} Homepage Component
 */
export function HomePage() {
  return (
    <PageWrapper>
      <OverlapGroup3>
        <OverlapGroup1>
          <BlobShape1>
            <img loading="lazy" src={Constants.BLOB_PNG_URL} alt="IMG" />
          </BlobShape1>
        </OverlapGroup1>
        <Heading>
          <Title>Need Groceries?</Title>
          <Description>
            Buy Groceries Quick, and get them even quicker!
          </Description>
        </Heading>
        <OverlapGroup2>
          <OAuthButton />
        </OverlapGroup2>
      </OverlapGroup3>
    </PageWrapper>
  );
}

const OverlapGroup3 = styled.div`
  width: 100%;
  height: 80%;
  position: relative;
  margin-left: 3%;
`;

const OverlapGroup1 = styled.div`
  width: 100%;
  top: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  padding: 3% 0;
  align-items: flex-start;
  min-height: 80%;
`;

const BlobShape1 = styled.div`
  width: 48%;
  align-self: flex-end;
  margin-top: 5%;
  margin-right: 3%;
  display: flex;
  align-items: flex-start;
  overflow: hidden;
  selection-color: blue;
`;

const Heading = styled.div`
  position: absolute;
  width: 50%;
  height: 20%;
  top: 30%;
  left: 0;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const Title = styled.h1`
  min-height: 50%;
  font-weight: 400;
  color: ${p => p.theme.grey};
  font-size: ${StyledConstants.FONT_SIZE_XXL};
  letter-spacing: 0;
  line-height: 100%;
  white-space: nowrap;
`;

const Description = styled.div`
  min-height: 40%;
  margin-top: 5;
  font-weight: 400;
  color: #66645e;
  font-size: ${StyledConstants.FONT_SIZE_L};
  letter-spacing: 0;
  line-height: 5%;
`;

const OverlapGroup2 = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 60%;
  left: 0;
`;
