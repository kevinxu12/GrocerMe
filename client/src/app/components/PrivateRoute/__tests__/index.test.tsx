
import * as React from 'react';
import { initialState } from 'store/auth/reducer';
import { HomePage } from 'app/pages/Landing';
import PrivateRoute from '..';
import { generateAuthState, MockStoreWrapper } from 'utils/test';
import { RootState } from 'types';
import * as authHelper from 'utils/auth';

jest.mock('utils/auth');
describe('<PrivateRoute  />', () => {
    const mockStoreWrapper = new MockStoreWrapper<RootState>({auth: initialState});
    it('Not logged in user should reroute to the home page', () => {
      const auth = initialState;
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<PrivateRoute auth = {auth} component = {HomePage} />, {auth});
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(authHelper.isPathAuthenticated).toBeCalledTimes(3); // once on render, once on useEffect, once on re-render
    });

    it('Logged in user should stay at the current page', () => {
      const auth = generateAuthState([]);
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<PrivateRoute auth = {auth} component = {HomePage} />, {auth});
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(authHelper.isPathAuthenticated).toBeCalledTimes(3);
    });
  });