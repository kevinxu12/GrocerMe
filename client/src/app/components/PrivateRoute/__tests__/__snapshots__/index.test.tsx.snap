// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateRoute  /> Logged in user should stay at the current page 1`] = `null`;

exports[`<PrivateRoute  /> Not logged in user should reroute to the home page 1`] = `null`;
